src/Cassandra/Cluster.cs

```diff
@@ -242,13 +244,32 @@ public async Task<ISession> ConnectAsync(string keyspace)
         async Task<TSession> IInternalCluster.ConnectAsync<TSession>(ISessionFactory<TSession> sessionFactory, string keyspace)
         {
             await Init().ConfigureAwait(false);
-            var session = await sessionFactory.CreateSessionAsync(keyspace, _serializer).ConfigureAwait(false);
+            var newSessionName = GetNewSessionName();
+            var session = await sessionFactory.CreateSessionAsync(keyspace, _serializer, newSessionName).ConfigureAwait(false);
             await session.Init().ConfigureAwait(false);
             _connectedSessions.Add(session);
             Cluster.Logger.Info("Session connected ({0})", session.GetHashCode());
             return session;
         }
 
+        private string GetNewSessionName()
+        {
+            var sessionCounter = GetAndIncrementSessionCounter();
+            if (sessionCounter == 0 && Configuration.SessionName != null)
+            {
+                return Configuration.SessionName;
+            }
+
+            var prefix = Configuration.SessionName ?? Configuration.DefaultSessionName;
+            return prefix + sessionCounter;
+        }
+
+        private long GetAndIncrementSessionCounter()
+        {
+            var newCounter = Interlocked.Increment(ref _sessionCounter);
```

Comment:
Is this because of overflow?
If so, we could add a code comment.

Reply:
Yeah just to avoid negative counters if it overflows, I'll add a comment